### PR TITLE
Specially handle some scenarios for type parameters with 'allows ref struct' constraint

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs [hunk 1]
```diff
@@ -527,9 +527,14 @@ public static bool IsPossibleArrayGenericInterface(this TypeSymbol type)
             return false;
         }
 
-        internal static bool IsErrorTypeOrRefLikeType(this TypeSymbol type)
+        internal static bool IsErrorTypeOrIsRefLikeTypeOrAllowByRefLike(this TypeSymbol type)
```
Comment (jjonescz) [inline src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs:530]:
```suggestion
        internal static bool IsErrorTypeOrIsRefLikeTypeOrAllowsByRefLike(this TypeSymbol type)
```

File: src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs [hunk 2]
```diff
@@ -16822,7 +16822,814 @@ public class Helper
...
+                Diagnostic(ErrorCode.ERR_ScopedRefAndRefStructOnly, "scoped T z").WithLocation(13, 26),
+                // (19,23): error CS9048: The 'scoped' modifier can be used for refs and ref struct values only.
+                //         var d = void (scoped T u) => {};   
+                Diagnostic(ErrorCode.ERR_ScopedRefAndRefStructOnly, "scoped T u").WithLocation(19, 23)
+                );
+
+            var src2 = @"
+public class Helper
+{
+    public static void Test1<T>(scoped T x)
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs:17019]: Consider testing `scoped ref T` parameter as well.

File: src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs [hunk 3]
```diff
@@ -16822,7 +16822,814 @@ public class Helper
...
+            void validate(ModuleSymbol m)
+            {
+                var p = m.GlobalNamespace.GetMember<MethodSymbol>("Helper.Test1").Parameters[0];
+                AssertEx.Equal("scoped T x", p.ToTestDisplayString());
+                Assert.Equal(ScopedKind.ScopedValue, p.EffectiveScope);
+            }
+        }
+
+        [Fact]
+        public void ScopedTypeParameter_02()
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs:17040]: Nit: looks like this is testing locals, not parameters.

File: src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs [hunk 4]
```diff
@@ -16822,7 +16822,814 @@ public class Helper
...
+        public void UnscopedRef_02()
+        {
+            var sourceA =
+@"
+using System.Diagnostics.CodeAnalysis;
+using System.Collections.Generic;
+
+public class Helper
+{
+    static void Test1<T>([UnscopedRef] params scoped T x)
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs:17223]: Consider testing similar scenario without `params`

File: src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs [hunk 5]
```diff
@@ -16822,7 +16822,814 @@ public class Helper
...
+                //     public static T P3;
+                Diagnostic(ErrorCode.ERR_FieldAutoPropCantBeByRefLike, "T").WithArguments("T").WithLocation(14, 19),
+                // (15,19): error CS8345: Field or auto-implemented property cannot be of type 'S' unless it is an instance member of a ref struct.
+                //     public static S P4;
+                Diagnostic(ErrorCode.ERR_FieldAutoPropCantBeByRefLike, "S").WithArguments("S").WithLocation(15, 19)
+                );
+        }
+
+        [Fact]
+        public void AutoProperty()
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs:17404]: Consider testing similar scenario with synthesized record properties

File: src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs [hunk 6]
```diff
@@ -16822,7 +16822,814 @@ public class Helper
...
+                );
+        }
+
+        [Fact]
+        public void MissingScopedInOverride_01()
+        {
+            var src = @"
+abstract class Base
+{
+    protected abstract T Test1<T>(scoped T x)
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs:17523]: Should we test similar scenario with `params T` which is implicitly scoped?

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs [hunk 7]
```diff
@@ -333,7 +333,7 @@ private enum IsParamsValues : byte
                         Debug.Assert(refKind != RefKind.None);
                         scope = ScopedKind.ScopedRef;
```
Comment (jjonescz) [inline src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs:334]: Can we have `T: allows ref struct` parameter symbol which is also byref, in metadata?

File: src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs [hunk 8]
```diff
@@ -16822,7 +16822,814 @@ public class Helper
...
+                Diagnostic(ErrorCode.ERR_ScopedMismatchInParameterOfOverrideOrImplementation, "Test2").WithArguments("y").WithLocation(14, 38)
+                );
+        }
+
+        [Fact(Skip = "'byreflike' in IL is not supported yet")] // PROTOTYPE(RefStructInterfaces): Enable once we get support for 'byreflike' in IL.
+        public void RefFieldTypeAllowsRefLike()
+        {
+            // ref struct R2<T> where T : allows ref struct
+            // {
+            //     public ref T F;
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs:17606]: Consider adding a test like this for a parameter as well.

**Comment by AlekseyTs** [comment]
@jjonescz  Thank you for the test suggestions. However, I am going through specific code paths (marked with PROTOTYPE comments, quate possibly some are marked in the first commit of this PR), adjusting the code paths (if necessary) and adding targeted tests for those specific code paths. I will capture you suggestions as PROTOTYPE comments in the next PR, but I will not be adding the tests at the moment.

File: src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs:530]: I prefer to do this rename in the next PR.

File: src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs (hunk 7 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Metadata/PE/PEParameterSymbol.cs:334]: > Can we have `T: allows ref struct` parameter symbol which is also byref, in metadata?

We can have one in source as well. There is nothing wrong with that.

**Review by jjonescz — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@cston, @dotnet/roslyn-compiler For the second review

File: src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs [hunk 9]
```diff
@@ -16822,7 +16822,814 @@ public class Helper
...
+                //     public S P2 {get; set;}
+                Diagnostic(ErrorCode.ERR_FieldAutoPropCantBeByRefLike, "S").WithArguments("S").WithLocation(6, 12),
+                // (12,19): error CS8345: Field or auto-implemented property cannot be of type 'T' unless it is an instance member of a ref struct.
+                //     public static T P3 {get; set;}
+                Diagnostic(ErrorCode.ERR_FieldAutoPropCantBeByRefLike, "T").WithArguments("T").WithLocation(12, 19),
+                // (13,19): error CS8345: Field or auto-implemented property cannot be of type 'S' unless it is an instance member of a ref struct.
+                //     public static S P4 {get; set;}
+                Diagnostic(ErrorCode.ERR_FieldAutoPropCantBeByRefLike, "S").WithArguments("S").WithLocation(13, 19)
+                );
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs:17448]: >} [](http://example.com/codeflow?start=8&length=1)

Consider testing capturing primary constructor parameter of type `allows ref struct`.

File: src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs [hunk 10]
```diff
@@ -16822,7 +16822,814 @@ public class Helper
...
+            var comp = CreateCompilation(src, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics);
+            comp.VerifyDiagnostics(
+                // (12,38): error CS8987: The 'scoped' modifier of parameter 'x' doesn't match overridden or implemented member.
+                //     protected abstract override void Test1<T>(T x, out T z);
+                Diagnostic(ErrorCode.ERR_ScopedMismatchInParameterOfOverrideOrImplementation, "Test1").WithArguments("x").WithLocation(12, 38),
+                // (14,38): error CS8987: The 'scoped' modifier of parameter 'y' doesn't match overridden or implemented member.
+                //     protected abstract override void Test2(S y, out S z);
+                Diagnostic(ErrorCode.ERR_ScopedMismatchInParameterOfOverrideOrImplementation, "Test2").WithArguments("y").WithLocation(14, 38)
+                );
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs:17599]: >} [](http://example.com/codeflow?start=8&length=1)

Consider testing `ERR_ScopedMismatchInParameterOfPartial` and `ERR_ScopedMismatchInParameterOfTarget`.

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs (hunk 8 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs:17606]: > Consider adding a test like this for a parameter as well.

Parameters like that are supported and there is already a bunch of tests covering them.

File: src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs (hunk 5 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs:17404]: > Consider testing similar scenario with synthesized record properties

`RestrictedTypesInRecords` will be added in the next PR

File: src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit3/RefStructInterfacesTests.cs:17019]: > Consider testing `scoped ref T` parameter as well.

Why do you think this is interesting? Any `ref` parameter can be declared scoped, no type restrictions for that.

**Pull request merged**